native/dispatch.c

```diff
@@ -3342,7 +3371,7 @@ Java_com_sun_jna_Native_registerMethod(JNIEnv *env, jclass UNUSED(ncls),
   if (closure_types) (*env)->ReleaseLongArrayElements(env, closure_atypes, closure_types, 0);
   if (cvts) (*env)->ReleaseIntArrayElements(env, conversions, cvts, 0);
   data->fptr = L2A(function);
-  data->closure_rclass = (*env)->NewWeakGlobalRef(env, closure_rclass);
+  data->closure_method = (*env)->NewGlobalRef(env, closure_method);
```

Comment:
Does it not work for you with a weak global?  The return type class will be referenced by the callback class, although I'm not sure how the JVM manages references to Method objects; it's quite possible Class objects don't actually have _any_ references to Method objects and they're only created on demand.